core/irCheck.ml

```diff
@@ -745,6 +745,14 @@ struct
 
               Query (range, e, t), t, o
 
+        | InsertRows (_source, _rows) ->
+	    (* TODO: check that the type of rows being inserted matches the table type *)
+	    assert false
```

Comment:
@frank-emrich, I haven't attempted to add code to typecheck the insertion operators.  I can take a shot at this and if this is straightforward, it would be good to do before merging, if not, I'll create an issue.

Reply:
I'm happy to create code for what I think should be happening here, so we can have a look at it together. I'd just commit it to this branch.

Reply:
Thanks!

Reply:
I've pushed a commit dealing with this. Do you have an idea which of ours test cases or example has fancy polymorphic inserts? I would like to test this.

If it's easier to discuss in person, I'm happy to come around.

Reply:
Thanks. I'm heading out shortly.  The code looks reasonable to me though.

As far as I know the only tests that exercise `insert` are in `tests/database/factorials.links`.  So running this with ir typing temporarily enabled should exercise this code.  In this PR I added some tests that exercise different syntaxes for inserts as well as `returning` clauses, to make sure that my changes work at least in common cases.  I'll be happy if IR typing works on these cases for now; if you want to add some additional examples to `factorials.links` that exercise the full generality of the IR typing code that would also be great.

Since we are already generating plenty of ill-typed IR, I'd rather get this merged and create an issue to revisit this later if there are obstacles to getting IR typing working for the specific case of inserts.

Reply:
OK. So actually there look like there are several problems with typechecking the insert/delete/update operations (in addition to lots of other IR checking issues).  For example:
```
--------------------------------------------------------------------
Continuing after IR Type Error:
Internal error: attempt to extract a row from a datatype that is not a record or a variant: [(f:Int,i:Int)]
noccurring in IR special tail computation:
(Ir.InsertRows ((Ir.Variable 2066),
   (Ir.TApp ((Ir.Variable 56), [(f:Int,i:Int)]))))
-------------------------------------------------------------------
```
I think this is because in the line 
```
            let rows_r = TypeUtils.extract_row rows_t in

```
you first need to get rid of the list constructor on rows_t since it is of type list of record.

You can see the errors by running with `typecheck_ir` and `debug` enabled, without any database configuration, the IR typing messages will get printed out before run-time database errors.

Reply:
Thanks for letting me know. I'll play around with this a bit and fix the bugs in this new fragment immediately. I'll fix the bugs in existing code independently from this PR, to untangle the two things.

Reply:
Thanks, seems to work now.